posthog-core/src/index.ts

```diff
@@ -672,39 +672,51 @@ export abstract class PostHogCoreStateless {
   }
 
   async shutdown(shutdownTimeoutMs: number = 30000): Promise<void> {
-    await this._initPromise
+    // A little tricky - we want to have a max shutdown time and enforce it, even if that means we have some
+    // dangling promises. We'll keep track of the timeout and resolve/reject based on that.
 
+    await this._initPromise
+    let hasTimedOut = false
     this.clearFlushTimer()
 
-    try {
-      await Promise.all(Object.values(this.pendingPromises))
-
-      const startTimeWithDelay = Date.now() + shutdownTimeoutMs
+    const doShutdown = async (): Promise<void> => {
+      try {
+        await Promise.all(Object.values(this.pendingPromises))
 
-      while (true) {
-        const queue = this.getPersistedProperty<PostHogQueueItem[]>(PostHogPersistedProperty.Queue) || []
+        while (true) {
+          const queue = this.getPersistedProperty<PostHogQueueItem[]>(PostHogPersistedProperty.Queue) || []
 
-        if (queue.length === 0) {
-          break
-        }
+          if (queue.length === 0) {
+            break
+          }
 
-        // flush again to make sure we send all events, some of which might've been added
-        // while we were waiting for the pending promises to resolve
-        // For example, see sendFeatureFlags in posthog-node/src/posthog-node.ts::capture
-        await this.flush()
+          // flush again to make sure we send all events, some of which might've been added
+          // while we were waiting for the pending promises to resolve
+          // For example, see sendFeatureFlags in posthog-node/src/posthog-node.ts::capture
+          await this.flush()
 
-        // If we've been waiting for more than the shutdownTimeoutMs, stop it
-        const now = Date.now()
-        if (startTimeWithDelay < now) {
-          break
+          if (hasTimedOut) {
+            break
+          }
         }
+      } catch (e) {
+        if (!isPostHogFetchError(e)) {
+          throw e
+        }
+        this.logMsgIfDebug(() => console.error('Error while shutting down PostHog', e))
       }
-    } catch (e) {
-      if (!isPostHogFetchError(e)) {
-        throw e
-      }
-      this.logMsgIfDebug(() => console.error('Error while shutting down PostHog', e))
     }
+
+    return Promise.race([
```

Comment:
> Shutdown will loop over and flush data but outside of issues with our ingestion endpoint I can't see why it would hang

Yeah same, I did not expect this happening but now it will bail out regardless.